lib/logger/lib/logger.ex

```diff
@@ -691,49 +700,54 @@ defmodule Logger do
           :ok | {:error, :noproc} | {:error, term}
   def bare_log(level, chardata_or_fun, metadata \\ []) do
     case __should_log__(level) do
-      :error -> :ok
-      info -> __do_log__(info, chardata_or_fun, metadata)
+      nil -> :ok
+      level -> __do_log__(level, chardata_or_fun, Map.new(metadata))
     end
   end
 
   @doc false
-  def __should_log__(level) when level in @levels do
-    case __metadata__() do
-      {true, pdict} ->
-        case Logger.Config.log_data(level) do
-          {:discard, _config} -> :error
-          {mode, config} -> {level, mode, config, pdict}
-        end
+  def __should_log__(:warn) do
+    # TODO: Uncomment this
+    # bare_log(:warning, "`:warn` level is deprecated, you should use `:warning`")
+
+    __should_log__(:warning)
+  end
 
-      {false, _} ->
-        :error
+  def __should_log__(level) when level in @levels do
+    if enabled?(self()) and not match?({:discard, _}, Logger.Config.log_data(level)) do
+      level
     end
   end
 
   @doc false
-  def __do_log__({level, mode, config, pdict}, chardata_or_fun, metadata)
-      when is_list(metadata) do
-    %{utc_log: utc_log?, truncate: truncate} = config
-    metadata = [pid: self()] ++ into_metadata(metadata, pdict)
-
-    case normalize_message(chardata_or_fun, metadata) do
-      {message, metadata} ->
-        tuple = {Logger, truncate(message, truncate), Logger.Utils.timestamp(utc_log?), metadata}
-
-        try do
-          notify(mode, {level, Process.group_leader(), tuple})
-          :ok
-        rescue
-          ArgumentError -> {:error, :noproc}
-        catch
-          :exit, reason -> {:error, reason}
-        end
+  def __do_log__(:warn, chardata_or_fun, metadata) do
+    __do_log__(:warning, chardata_or_fun, metadata)
+  end
 
+  def __do_log__(level, fun, metadata) when is_function(fun, 0) and is_map(metadata) do
+    case fun.() do
       :skip ->
         :ok
+
+      {msg, meta} ->
+        :logger.log(level, msg, Map.merge(metadata, Map.new(meta)))
```

Comment:
```suggestion
        :logger.log(level, msg, Enum.into(meta, metadata))
```

It should be the most efficient unless meta is really large.